drivers/sms_ser.c

```diff
@@ -0,0 +1,698 @@
+    }
+
+    if (!strcasecmp(cmdname, "test.battery.stop")) {
+        length = sms_prepare_cancel_test(&bufOut[0]);
+
+        if (ser_send_buf(upsfd, bufOut, length) == 0) {
+            upsdebugx(3, "failed to send test.battery.stop");
+            return STAT_INSTCMD_FAILED;
+        }
+        upsdebugx(3, "command test.battery.stop OK!");
+
+        return STAT_INSTCMD_HANDLED;
+    }
+
+    if (!strcasecmp(cmdname, "beeper.toggle")) {
+        length = sms_prepare_set_beep(&bufOut[0]);
+
+        if (ser_send_buf(upsfd, bufOut, length) == 0) {
+            upsdebugx(3, "failed to send beeper.toggle");
+            return STAT_INSTCMD_FAILED;
+        }
+        upsdebugx(3, "command beeper.toggle OK!");
+
+        return STAT_INSTCMD_HANDLED;
+    }
+
+    if (!strcasecmp(cmdname, "shutdown.return")) {
+        length = sms_prepare_shutdown_restore(&bufOut[0]);
+
+        if (ser_send_buf(upsfd, bufOut, length) == 0) {
+            upsdebugx(3, "failed to send shutdown.return");
+            return STAT_INSTCMD_FAILED;
+        }
+        upsdebugx(3, "command shutdown.return OK!");
+
+        return STAT_INSTCMD_HANDLED;
+    }
+
+    if (!strcasecmp(cmdname, "shutdown.reboot")) {
+        long delay = extra ? strtol(extra, NULL, bootdelay) : bootdelay;
+        length = sms_prepare_shutdown_nsec(&bufOut[0], delay);
+
+        if (ser_send_buf(upsfd, bufOut, length) == 0) {
+            upsdebugx(3, "failed to send shutdown.reboot");
+            return STAT_INSTCMD_FAILED;
+        }
+        upsdebugx(3, "command shutdown.reboot OK!");
+
+        return STAT_INSTCMD_HANDLED;
+    }
+
+    if (!strcasecmp(cmdname, "shutdown.stop")) {
+        length = sms_prepare_cancel_shutdown(&bufOut[0]);
+
+        if (ser_send_buf(upsfd, bufOut, length) == 0) {
+            upsdebugx(3, "failed to send shutdown.stop");
+            return STAT_INSTCMD_FAILED;
+        }
+        upsdebugx(3, "command shutdown.stop OK!");
+
+        return STAT_INSTCMD_HANDLED;
+    }
+
+    upslogx(LOG_NOTICE, "sms_instcmd: unknown command [%s]", cmdname);
+    return STAT_INSTCMD_UNKNOWN;
+}
+
+static int sms_setvar(const char *varname, const char *val) {
+    if (!strcasecmp(varname, "ups.delay.reboot")) {
+        int ipv = atoi(val);
+        if (ipv >= 0)
+            bootdelay = (unsigned int)ipv;
+        dstate_setinfo("ups.delay.reboot", "%u", bootdelay);
+        return STAT_SET_HANDLED;
+    }
+    return STAT_SET_UNKNOWN;
+}
+
+void upsdrv_initinfo(void) {
+    upslogx(LOG_INFO, "upsdrv_initinfo");
+    char *battery_status;
+
+    if (get_ups_features() != 0) {
+        upslogx(LOG_ERR, "Short read from UPS");
+        dstate_datastale();
+        return;
+    }
+    if (get_ups_information() != 0) {
+        upslogx(LOG_ERR, "Short read from UPS");
+        dstate_datastale();
+        return;
+    }
+
+    if (get_ups_nominal() == 0) {
+        dstate_setinfo("device.model", "%s", (unsigned char *)DeviceData.model);
```

Comment:
Was casting to `(unsigned char *)` required? I think `%s` deals with `char*` just fine...

Reply:
Thats a good question! i think that it not necessary to do this cast. i will fix.